src/linux/functions.cpp

```diff
@@ -15,27 +15,70 @@
  */
 
 #include <sys/mount.h>
+#include <mntent.h>
 #include <errno.h>
 #include "../mountutils.hpp"
 
 NAN_METHOD(UnmountDisk) {
   if (!info[1]->IsFunction()) {
-    return Nan::ThrowError("Callback must be a function");
+    return Nan::ThrowTypeError("Callback must be a function");
   }
 
   v8::Local<v8::Function> callback = info[1].As<v8::Function>();
 
   if (!info[0]->IsString()) {
-    return Nan::ThrowError("Device argument must be a string");
+    return Nan::ThrowTypeError("Device argument must be a string");
   }
 
   v8::String::Utf8Value device(info[0]->ToString());
-  int code = umount(reinterpret_cast<char *>(*device));
 
-  if (code == 0) {
-    YIELD_NOTHING(callback);
+  const char *device_path = reinterpret_cast<char *>(*device);
+  const char *mount_path = NULL;
+
+  // Get mountpaths from the device path, as `umount(device)`
+  // has been removed in Linux 2.3+
+  struct mntent *mount_entity;
+  FILE *proc_mounts;
+
+  proc_mounts = setmntent("/proc/mounts", "r");
+
+  if (proc_mounts == NULL) {
+    v8::Local<v8::Value> argv[1] = {
+      Nan::ErrnoException(errno, "setmntent", NULL, "/proc/mounts")
+    };
+    Nan::MakeCallback(Nan::GetCurrentContext()->Global(), callback, 1, argv);
+    return;
+  }
+
+  // TODO(jhermsmeier): Support únmounting multiple mountpoints (?)
+  // NOTE: Might not be necessary, as MNT_DETACH will cause
+  // every mount in the mount namespace to be lazily unmounted
+  while ((mount_entity = getmntent(proc_mounts)) != NULL) {
+    mount_path = mount_entity->mnt_dir;
+    if (strncmp(mount_path, device_path, strlen(device_path)) == 0) {
+      break;
+    }
+  }
+
+  endmntent(proc_mounts);
+
+  if (mount_path == NULL) {
+    v8::Local<v8::Value> argv[1] = { Nan::Error("Couldn't find mount paths") };
+    Nan::MakeCallback(Nan::GetCurrentContext()->Global(), callback, 1, argv);
+    return;
+  }
+
+  // Use umount2() with the MNT_DETACH flag, which performs a lazy unmount;
+  // makíng the mount point unavailable for new accesses,
+  // and only actually unmounting when the mount point ceases to be busy
```

Comment:
> call to determine when the drive is really unmounted?

Afaik, it won't return until the operation is complete, so once it returns it'll all have happened already, and there's no need to guard against async stuff where there is none.

> given that we're only unmounting in order to be able to write a raw disk image to the drive, we probably don't need to care about any unflushed write buffers

As mentioned below (https://github.com/resin-io-modules/mountutils/pull/10#discussion_r106496933); this can have system-wide effects, i.e. when you're not the only process on the system accessing a mount or file connected to it